torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5119,3 +5119,30 @@ def test_output_unused_in_loss(self):
                                 )
                         else:
                             self.assertFalse(True, "DDP error not raised")
+
+        @require_backend({"gloo",})
+        @require_backends_available({"gloo",})
+        def test_monitored_barrier_gloo(self):
+            process_group = dist.new_group(ranks=list(i for i in range(int(self.world_size))))
+            tensors = [torch.ones(10) * self.rank]
+            # Kick off some allreduce work on all ranks
+            for _ in range(10):
+                process_group.allreduce(tensors).wait()
+            # Run monitored barrier
+            timeout = 5
+            process_group.monitored_barrier(timeout)
+            # All ranks besides 1 call into barrier, rank 0 should report failure
+            # while others report gloo error.
+            failed_rank = 1
+            if self.rank == failed_rank:
+                return
+            if self.rank == 0:
+                with self.assertRaisesRegex(RuntimeError, f"Rank {failed_rank}"):
+                    process_group.monitored_barrier(timeout)
+            else:
+                # Other ranks will report standard gloo error, only rank 0 knows
```

Comment:
what if all reduce has error in rank 0?

Reply:
The main use case here is to report hangs/issues for non-zero ranks. Most users read master (rank 0) logs first, so issues on rank 0 should be pretty apparent, the challenge is usually hangs in nonzero ranks. 

I think one small enhancement here could be to report a better error when nonzeroro ranks timeout on send/recv (currently we just let it crash with gloo error), to indicate something may have been crashed in rank 0.